erts/emulator/beam/erl_gc.c

```diff
@@ -562,7 +562,10 @@ delay_garbage_collection(Process *p, int need, int fcalls)
 
         /* Point at the end of the address range to ensure that
          * test for the safe range in the new heap in the
-         * update_record_in_place instruction fails. */
+         * update_record_in_place instruction fails. Save the previous
+         * value of the high water mark to make it possible to make the
+         * next GC a minor one. */
+        p->abandoned_high_water = p->high_water;
```

Comment:
Would another way be to check `p->abandoned_heap == NULL` in the `update_record_in_place` instruction? It would add a little bit of overhead there, but save memory in the process struct.

Reply:
Yes, that would be possible. I will have to think about that. Another way could be to only overwrite `p->high_water` if the JIT is enabled. That would save memory for anyone that does not use the JIT.